Add a deferred call hardware abstraction and multiplexer

lschuermann opened the pull request:
### Pull Request Overview
To build truly hardware-independent applications with Tock, it may be feasible to emulate some peripherals in the kernel. This could be especially useful for algorithms like SHA or AES: If a crypto-accelerator is present, it may be used; if not, a software implementation can fill the gap.

For this to work, software devices need to behave similar to hardware devices. This means, that callbacks need to work as if they've been triggered by an interrupt.

This pull request adds a hardware abstraction over deferred calls and makes them easily and safely accessible to all kinds of kernel code, like capsules.


### Testing Strategy

This pull request was tested by using the `DeferredCallMux` in a prototype `SHA2-32` implementation. Further testing is required.


### TODO or Help Wanted

As this is my first PR for Tock, I'd like to receive general feedback. There probably is a lot of room for improvement.


### Documentation Updated

- [ ] Updated the relevant files in `/docs`, or no updates are required.

### Formatting

- [x] Ran `make formatall`.

bradjc commented:
I think this is pretty awesome, and as you point out, something we are likely to need soon.

I'm wondering if there is some way to remove the need for `MUXBACKEND`, and basically make the mux a more integrated component of the already kernel based `DeferredCall` mechanism. Something like making bit 31 of `DEFERRED_CALL` the dedicated "capsule deferred call bit". The `chip.rs` interrupt handling loop would still need to know that the bit was set so that it could choose how to schedule it, but it wouldn't need to define anything in the chip crate. It could instead call back to the kernel library and let it call the correct client code.

I'm not sure I have thought through all of the details, but it would make this easier to use for new/other chips.

lschuermann commented on kernel/src/common/deferred_call_mux.rs:
```diff
@@ -0,0 +1,207 @@
+    // If the returned reference is identical to the mux argument,
+    // it is set in the option. Otherwise, a different mux is
+    // already registered and may not be replaced.
+    (*DEFERRED_CALL_MUX.get_or_insert(mux)) as *const _ == mux as *const _
```
I'm not quite sure whether it's a good idea to prevent setting a global `Mux` a second time. While this could result in a `Mux` getting called even though it hasn't request a deferred call, nothing too bad should happen if the global Mux gets changed.

niklasad1 replied:
I don't understand why you would replace it or need to configure it explicitly in separate method instead of doing this in the constructor. 

However, the docs are really good :+1:

lschuermann replied:
I don't know how it would be possible for me to configure the global `Mux` reference in the constructor, as I can't create a reference that persists while the struct is still being moved to it's final destination, right?

I assumed that it is best practice to `static_init!` only in the `boards`-crate, where all other capsules are initialized too. So the `set_global_mux` function serves a somewhat similar purpose as all the `set_client`s.

niklasad1 replied:
Right, I think I got confused and mixed up `DeferredCallMux` and `DeferredCallMuxClient` sorry!

lschuermann replied:
No worries. I think my naming is a bit confusing and way to _enterprisified_. In general I'm having a hard time coming up with good names for all the structs, so suggestions are welcome. :laughing:

lschuermann commented on kernel/src/common/deferred_call_mux.rs:
```diff
@@ -0,0 +1,207 @@
+/// This function needs to be called by the underlying deferred
+/// call implementation in the `chip` crate.
+pub unsafe fn call_global_mux() -> bool {
+    DEFERRED_CALL_MUX.map(|mux| mux.call()).is_some()
```
Because Tock's design is limited to single threaded code execution, this unsafe part shouldn't cause too much trouble, right?

niklasad1 replied:
I would claim that the most important thing as such executions can't be interleaved e.g, interrupts in `not-reentrant` functions such as this one, but because the kernel is event-driven (run-to-completion) it should be safe AFAIK.

However, `tock` more or less forces this design unless you manually implement `unsafe Send + Sync`

/cc @alevy please correct me if I'm wrong

lschuermann commented on kernel/src/common/deferred_call_mux.rs:
```diff
@@ -0,0 +1,207 @@
+//!
+//! let deferred_call_mux_clients = unsafe { static_init!(
+//!     [(Cell<bool>, OptionalCell<&'static DeferredCallMuxClient>); 1],
+//!     [(Cell::new(false), OptionalCell::empty())]
```
Initializing the internal data structure separately seems rather inelegant. Is there some more idiomatic way to do this?

niklasad1 replied:
You could maybe replace the `tuple (Cell, OptionCell)` with a struct and impl `Default` for it.

However, personally I think `OptionCell` should impl `Default` as `None` then you could `Default::default()` for them

lschuermann replied:
> You could maybe replace the `tuple (Cell, OptionCell)` with a struct and impl `Default` for it.

Done.

> However, personally I think `OptionCell` should impl `Default` as `None` then you could `Default::default()` for them

Sounds like a great idea. I've seen you already created a PR for that. :+1:

lschuermann commented on kernel/src/common/deferred_call_mux.rs:
```diff
@@ -0,0 +1,207 @@
+    /// It may be called without holding the `DeferredCallMux` reference through
+    /// `call_global_mux`.
+    pub(self) fn call(&self) {
+        self.clients
```
Would it make sense to round-robin / shuffle call the deferred calls?

lschuermann commented:
@bradjc I've removed the `MuxBackend`, but the `DeferredCallMux` still needs to be known in the `chip`'s deferred-call-tasks enum. However, I think that without the `Backend` and the highest bit reserved for `DeferredCallMux` calls, the whole design is more streamlined across different chips.

I'm still now sure whether the overall design makes sense, and I've added some specific questions to the PR. I'd love to hear feedback.

niklasad1 commented on kernel/src/common/deferred_call_mux.rs:
```diff
@@ -0,0 +1,207 @@
+    /// If no client for the handle is found (it was unregistered), this
+    /// returns `None`. If a call is already scheduled, it returns
+    /// `Some(false)`.
+    pub fn set(&self, handle: DeferredCallHandle) -> Option<bool> {
```
This seems fine to me but wouldn't it be better to return `Result<(), Error>` here and use `Error::AlreadySet` and `Error::ClientNotRegistered`?

Returning `Some(false)`  made it a bit hard for me grasp at least

lschuermann replied:
Yes, I also think a `Result` better suits this function. However, I would then need to introduce yet another error enum for the `Mux` (e.g. something like `DeferredCallMuxSetCallError`, which is a bit crazy right?).

Also, do you think that `AlreadySet` should be considered an error? The caller is scheduled in this case, so I'm not quite sure what to think about this. I also don't think it's a good idea to hide this information as setting the handle twice could indicate a logic error.

niklasad1 replied:
I think it is better to keep as it is for now, it is clean enough!

I think this belong in the conversation to create a  `chained result type` for all results in the kernel to encourage `Result` over `Option`

niklasad1 commented on kernel/src/common/deferred_call_mux.rs:
```diff
@@ -0,0 +1,207 @@
+use crate::common::deferred_call::DeferredCall;
+use core::cell::Cell;
+
+pub const DEFERRED_CALL_MUX_TASK: usize = 31; // highest bit in usize reserved for deferred call mux
```
make this comment `rust doc` instead?

lschuermann replied:
I've made it into a rustdoc comment. In addition to that, the `DEFERRED_CALL_MUX_TASK` is now always set to the highest bit in a usize, regardless of a 32- or 64-bit architecture.

niklasad1 commented on kernel/src/common/deferred_call_mux.rs:
```diff
@@ -0,0 +1,207 @@
+pub unsafe fn set_global_mux(mux: &'static DeferredCallMux) -> bool {
+    // If the returned reference is identical to the mux argument,
+    // it is set in the option. Otherwise, a different mux is
+    // already registered and may not be replaced.
```
```suggestion
    // already registered and will not be replaced.
```

niklasad1 commented on kernel/src/common/deferred_call_mux.rs:
```diff
@@ -0,0 +1,207 @@
+    ) -> DeferredCallMux {
+        DeferredCallMux {
+            deferred_call: DeferredCall::new(DEFERRED_CALL_MUX_TASK),
+            clients: clients,
```
```suggestion
            clients,
```

niklasad1 commented on kernel/src/common/deferred_call_mux.rs:
```diff
@@ -0,0 +1,207 @@
+/// is determined by the `clients`-array passed in with the constructor.
+pub struct DeferredCallMux {
+    deferred_call: DeferredCall<usize>,
+    clients: &'static [(Cell<bool>, OptionalCell<&'static DeferredCallMuxClient>)],
```
I'm not a big fan of `tuples` for readability purposes personally I would prefer separate struct for this but I guess this is simple enough or maybe a type alias?!

lschuermann replied:
I'll change this to be a struct and `impl Default` for it, so the clients-array can be easily initialized.

niklasad1 commented on kernel/src/common/deferred_call_mux.rs:
```diff
@@ -0,0 +1,207 @@
+    /// This function needs to be called by the underlying deferred call implementation.
+    /// It may be called without holding the `DeferredCallMux` reference through
+    /// `call_global_mux`.
+    pub(self) fn call(&self) {
```
why is `pub(self)` needed?

lschuermann replied:
The `call`-method should really only be called from inside this module, via the `call_global_mux` function. I don't see any case where further exposing this method would be helpful. That's why it's only exposed to the current module.

niklasad1 commented on kernel/src/common/deferred_call_mux.rs:
```diff
@@ -0,0 +1,207 @@
+    pub(self) fn call(&self) {
+        self.clients
+            .iter()
+            .map(|(ref call_reqd, ref oc)| (call_reqd, oc))
```
this seems map seems needless to me, `iter()` should be as_ref
```suggestion
```

lschuermann replied:
Yes, this `map` doesn't seem to be necessary. I've removed it in the `DeferredCallMuxClientState` commit.

bradjc commented:
These updates are excellent! After some offline discussion this PR is about 2 small steps from mergable.

1. Since really what this is adding is a dynamic deferred call feature, the name should be something like `DeferredCallDynamic`. Both the existing code and this code are acting as a mux, so noting that this version allows for things to be setup at runtime will help distinguish it from what is already there.

2. To avoid having to change every chip to support this, the check for whether there are any pending dynamic deferred calls should be moved to the main kernel loop (instead of being in the chip's interrupt checking loop). It should also be a lower priority than checking for actual interrupts. This would remove the need to reserve a special bit in the existing mux for this feature, and would mean that individual chip authors do not have to make sure (or decided whether or not) to include this functionality.

lschuermann commented:
Yes, I agree that it really should be called `Dynamic`, as that's what it really allows: dynamically registering `DeferredCall`. Should we go with `DeferredCallDynamic` or `DynamicDeferredCall`?

---

> 2. To avoid having to change every chip to support this, the check for whether there are any pending dynamic deferred calls should be moved to the main kernel loop

I get why this would be a good idea, however, not changing anything in the chip-crate would remove the chips ability to determine when a dynamic deferred call will be scheduled.
There's two options I could imagine here:
- Replace the `next_pending()` function with a `next_pending_or_call_dynamic()`, which as the name implies either returns the next _real_ deferred call, or if none is set, call the dynamic deferred calls.
- Add an additional `try_call_dynamic()`, which the chip-crate calls whenever there is no other interrupt / deferred call to handle. This however would have the risk of the chip-crate not implementing dynamic deferred calls.

lschuermann commented:
@bradjc I integrated the dynamic deferred calls into the main kernel loop. Is this how you intended it to be?

phil-levis commented:
Can I bump this? @bradjc can you respond?

lschuermann commented:
@bradjc Do you have any feedback on this?

I'd like to create RFCs/PRs for a software AES and SHA implementation (completely written in safe Rust as a capsule). These implementations heavily depend on dynamic deferred calls, and with uncertainty regarding naming etc., I'd rather avoid the merge chaos and rebase our implementations on a master with this PR integrated.

alevy commented:
@lschuermann this is way better! I think this is probably mergable as is, though I wonder if it would be better (and possible) to make the deferred call mux simply part of the `Kernel` struct rather than forcing the board to allocate is separately.

If there is a good reason not to do it, I'm fine with this as it is.

alevy commented on kernel/src/common/deferred_call_mux.rs:
```diff
@@ -0,0 +1,239 @@
+//! to set and receive deferred calls.
+//!
+//! These are scheduled "interrupts" in the chip scheduler.
+//! They are especially important if some hardware doesn't
```
We should remove the part about using this for hardware that doesn't support necessary interrupts, since we want that mechanism to be separate (and the responsibility of the chip implementation). Can we focus the comment of this module to just be on supporting asynchronous software operations (such as long running crypto operations)?

lschuermann replied:
Done. I hope my explanation for returning the call-stack up to the scheduler is clear enough. This is actually one of the primary use-cases for me, as my function call often happens in a `TakeCell`, and calling back to the client again tries to map that cell, failing of course.

lschuermann commented:
> I wonder if it would be better (and possible) to make the deferred call mux simply part of the `Kernel` struct

I think this could be a great improvement usability-wise, but the number of dynamic deferred calls depends on the board and capsules used. In my current implementation, I'm passing in an array of fixed size for the deferred call handles.
Maybe we could change this to be a macro or something? Though I'm not entirely sure on how to do this.

lschuermann commented:
I've implemented the requested changes.

> to make [it] part of the Kernel struct rather than forcing the board to allocate is separately
Is there any way around the requirements stated in my previous comment?

Is there something else that needs to be done before merging?
